### PR TITLE
mgr: fix metadata handling from old MDS daemons

### PR DESCRIPTION
**Pull request by jcsp**
Completely untested try at fixing the FIXME mentioned in https://github.com/ceph/ceph/pull/13617#issuecomment-288582611

Signed-off-by: John Spray <john.spray@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
IIRC the crash was on line 501 (when deciding whether to update) so by itself this patch won't help.  THis is intended to be used along with abde684?

**Comment by jcsp** [comment]
Apologies for long delay.

This patch should work, because the case on line 501 only applies when there is already some metadata in the map, and this change inserts it whenever we put it into the map.

**Comment by jcsp** [comment]
Oh, but the metadata could also get inserted the other way via load_all_metadata, in which case it would be missing the addr field.  So yes, this works in conjunction with the abde684 change -- this patch is still important to avoid repeatedly reloading the metadata for old daemons.

**Comment by jcsp** [comment]
Rebased

File: src/mgr/Mgr.cc [hunk 1]
```diff
@@ -94,6 +101,13 @@ class MetadataUpdate : public Context
 
         json_spirit::mObject daemon_meta = json_result.get_obj();
 
+        // Apply any defaults
+        for (const auto &i : defaults) {
```
Comment (tchaikov) [inline src/mgr/Mgr.cc]: `daemon_meta.insert(defaults.begin(), defaults.end());` would suffice.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-04-28_16:57:50-rados-wip-sage-testing2---basic-smithi/

**Pull request merged**